app/models/deck.rb

```diff
@@ -7,4 +7,31 @@ class Deck < ActiveRecord::Base
   validates :title, presence: true, uniqueness: {
               scope: :user_id
             }
+
+  def add_cards(cards)
+    deck_cards = cards.map do |name, quantity|
+      card = Card.find_by(name: name)
+      deck_card = DeckCard.new(card_id: card.id, quantity: quantity)
+    end
+    self.deck_cards = self.deck_cards + deck_cards
+  end
+
+  def sample_deck
+    self.cards.sample(7)
+  end
+
+  def card_type_curve
+    type_count = Hash.new
+    card_type = self.cards.card_type.to_a
```

Comment:
`self.cards.card_type` is a common mistake to make. `cards` is a has_many meaning that what you'll get back is an ActiveRecord relation containing a bunch of cards. You can't call `card_type` on a *relation* or *array of cards*, you can only call `card_type` on a *single* card.